File: src/main/java/core/procedures/io/PeekChar.java
package core.procedures.io;

import core.Repl;
import core.exceptions.SCMIOException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMInputPort;

import java.io.IOException;

public final class PeekChar extends AFn {

  public PeekChar() {
    super(new FnArgsBuilder().max(1).rest(SCMInputPort.class).build());
  }

  @Override
  public String getName() {
    return "peek-char";
  }

  @Override
  public Object apply(Object... args) {
    SCMInputPort inputPort;
    if (args.length == 0) {
      inputPort = Repl.getCurrentInputPort();
    } else {
      inputPort = ((SCMInputPort)args[0]);
    }
    try {
      return (char)inputPort.peek();
    } catch (IOException e) {
      throw new SCMIOException(e);
    }
  }
}

File: src/main/java/core/procedures/io/PeekChar.kt
package core.procedures.io

import core.Repl
import core.procedures.AFn
import core.procedures.Arity.Range
import core.scm.InputPort

class PeekChar : AFn<Any?, Char>(name = "peek-char", arity = Range(0, 1), restArgsType = InputPort::class.java) {

    override operator fun invoke(args: Array<out Any?>) = when (args.size) {
        0    -> Repl.currentInputPort
        else -> args[0] as InputPort
    }.peek().toChar()
}
